channeldb/migration25/migration.go

```diff
@@ -0,0 +1,211 @@
+package migration25
+
+import (
+	"bytes"
+	"fmt"
+
+	mig "github.com/lightningnetwork/lnd/channeldb/migration_01_to_11"
+	"github.com/lightningnetwork/lnd/kvdb"
+)
+
+var (
+	// openChanBucket stores all the currently open channels. This bucket
+	// has a second, nested bucket which is keyed by a node's ID. Within
+	// that node ID bucket, all attributes required to track, update, and
+	// close a channel are stored.
+	openChannelBucket = []byte("open-chan-bucket")
+
+	// chanInfoKey can be accessed within the bucket for a channel
+	// (identified by its chanPoint). This key stores all the static
+	// information for a channel which is decided at the end of  the
+	// funding flow.
+	chanInfoKey = []byte("chan-info-key")
+
+	// ErrNoChanDBExists is returned when a channel bucket hasn't been
+	// created.
+	ErrNoChanDBExists = fmt.Errorf("channel db has not yet been created")
+
+	// ErrNoActiveChannels  is returned when there is no active (open)
+	// channels within the database.
+	ErrNoActiveChannels = fmt.Errorf("no active channels exist")
+
+	// ErrChannelNotFound is returned when we attempt to locate a channel
+	// for a specific chain, but it is not found.
+	ErrChannelNotFound = fmt.Errorf("channel not found")
+)
+
+// MigrateInitialBalances patches the two new fields, InitialLocalBalance and
+// InitialRemoteBalance, for all the open channels. It does so by reading the
+// revocation log at height 0 to learn the initial balances and then updates
+// the channel's info.
+// The channel info is saved in the nested bucket which is accessible via
+// nodePub:chainHash:chanPoint. If any of the sub-buckets turns out to be nil,
+// we will log the error and continue to process the rest.
+func MigrateInitialBalances(tx kvdb.RwTx) error {
```

Comment:
(in the breach handling)

Reply:
yeah it's only used in rpc server to give the push amount info.

Reply:
The rpc-reported push can be wrong if it's dust, but nbd. It seems like the migration can be avoided by:
- persisting it for new channels w/o ChannelCommitment-based revlog (PR does this)
- existing channels call `fetchChanInfo`, tlv stream will end early and show that the initial balances weren't read (types not parsed with `DecodeWithParsedTypes`). Then just insert them by fetching the ChannelCommitment for height 0 since it _must_ store a ChannelCommitment in this case.

This is still technically a migration, but I think it's a lot less code

Reply:
I agree, would mean more code in the actually-used code path, but would have been easier for review. So nbd

Reply:
Have to say reviewing the migration-related code is not an easy task. I somewhat intentionally put some of the migration code here so we won't need to review a very large chunk of code in the following pr, tho it's already large...